building_sim_plugins/building_gazebo_plugins/src/crowd_simulator.cpp

```diff
@@ -0,0 +1,499 @@
+            << std::endl;
+      return false;
+    }
+
+    if (!modelTypeElement->HasElement("initial_pose"))
+    {
+      gzerr <<
+        "No model initial pose configured in <model_type>! <initial_pose> Required [" << s << "]" << std::endl;
+      return false;
+    }
+    if (!this->_LoadModelInitPose(modelTypeElement, modelTypePtr->pose))
+    {
+      gzerr <<
+        "Error loading model initial pose in <model_type>! Check <initial_pose> in [" << s << "]" << std::endl;
+      return false;
+    }
+
+    modelTypeElement = modelTypeElement->GetNextElement("model_type");
+  }
+
+  if (!sdf->HasElement("external_agent"))
+  {
+    gzwarn <<
+      "No external agent provided. <external_agent> is needed with a unique name defined above."
+           << std::endl;
+  }
+  auto externalAgentElement = sdf->GetElement("external_agent");
+  while (externalAgentElement)
+  {
+    auto exAgentName = externalAgentElement->Get<std::string>();
+    if (!this->_world->ModelByName(exAgentName))
+    {
+      gzerr << "Model " << exAgentName <<
+        " not defined in world! Check <external_agent>" << std::endl;
+      return false;
+    }
+    gzdbg << "Added external agent: [ " << exAgentName << " ]." << std::endl;
+    this->_externalAgents.emplace_back(exAgentName); //just store the name
+    externalAgentElement =
+      externalAgentElement->GetNextElement("external_agent");
+  }
+
+  return true;
+}
+
+//============================================
+bool CrowdSimulatorPlugin::_LoadCrowdSim()
+{
+  _crowdSimInterface = std::make_shared<crowd_simulator::CrowdSimInterface>(
+    _resourcePath,
+    _behaviorFile,
+    _sceneFile,
+    _simTimeStep);
+
+  assert(_crowdSimInterface);
+
+  //create crowd sim object database, loaded all agents defined in scene file
+  //assigned external agents with specific model name in world
+  //internal agents named by "agent[NO.]"
+  _crowdSimInterface->SpawnObject(this->_externalAgents);
+
+  //create model in world for each internal agents
+  size_t objectCount = this->_crowdSimInterface->GetNumObjects();
+  for (size_t id = 0; id < objectCount; ++id)
+  {
+    if (!this->_crowdSimInterface->GetObjectById(id)->isExternal)
+    {
+      auto objectPtr = this->_crowdSimInterface->GetObjectById(id);
+      assert(objectPtr);
+
+      //TODO: specify different modelType
+      // (size_t) agentPtr->_class identifies the profile type defined in scene.xml
+      // this is corresponds to the modelType id
+      auto typePtr = this->_modelTypeDBPtr->Get(objectPtr->typeName);
+      assert(typePtr);
+      if (!this->_CreateModel(objectPtr->modelName, typePtr,
+        objectPtr->agentPtr) )
+      {
+        gzerr << "Failed to insert model [ "<< objectPtr->modelName <<
+          " ] in world" << std::endl;
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+
+//============================================
+bool CrowdSimulatorPlugin::_LoadModelInitPose(
```

Comment:
Same as the above, can be moved to the common library?

Reply:
Yes, this can be done